### PR TITLE
Weapons are now marginally functional, can aim and shoot and stuff replicates over the network

### PR DESCRIPTION
**Pull request by kberg-amzn**
Weapons are now marginally functional, can aim and shoot and stuff replicates over the network

Signed-off-by: kberg-amzn <karlberg@amazon.com>

### CODE REVIEW

File: Gem/Code/Source/Components/NetworkWeaponsComponent.cpp [hunk 1]
```diff
@@ -111,61 +141,97 @@ namespace MultiplayerSample
...
+                {
+                    AZLOG_WARN("Failed to get transform for fire bone %s, joint Id %u", GetFireBoneNames(aznumeric_cast<uint32_t>(weaponIndex)).c_str(), m_fireBoneJointIds[weaponIndex]);
+                }
+
+                const FireParams&    fireParams = weapon->GetFireParams();
+                const AZ::Vector3    position = fireBoneTransform.GetTranslation();
+                const AZ::Quaternion orientation = AZ::Quaternion::CreateShortestArc(AZ::Vector3::CreateAxisX(), (fireParams.m_targetPosition - position).GetNormalized());
+                const AZ::Transform  transform = AZ::Transform::CreateFromQuaternionAndTranslation(orientation, position);
+                ActivateEvent activateEvent{ transform, fireParams.m_targetPosition, GetNetEntityId(), Multiplayer::InvalidNetEntityId };
+
```
Comment (AMZN-Gene) [inline Gem/Code/Source/Components/NetworkWeaponsComponent.cpp:198]: isReplay unused var

File: Gem/Code/Source/Weapons/IWeapon.h [hunk 2]
```diff
@@ -74,6 +75,30 @@ namespace MultiplayerSample
         //! @fireParams structure containing the internal fire params
         virtual void SetFireParams(const FireParams& fireParams) = 0;
 
+        //! Internally called on every weapon activation.
+        //! @param deltaTime              the time in seconds this activate event corresponds to
+        //! @param weaponState            the weapons internal state structure
+        //! @param weaponOwner            the weapons owning entity
```
Comment (AMZN-Gene) [inline Gem/Code/Source/Weapons/IWeapon.h:81]: weapon’s

File: Gem/Code/Source/Weapons/IWeapon.h [hunk 3]
```diff
@@ -74,6 +75,30 @@ namespace MultiplayerSample
         //! @fireParams structure containing the internal fire params
         virtual void SetFireParams(const FireParams& fireParams) = 0;
 
+        //! Internally called on every weapon activation.
+        //! @param deltaTime              the time in seconds this activate event corresponds to
+        //! @param weaponState            the weapons internal state structure
```
Comment (AMZN-Gene) [inline Gem/Code/Source/Weapons/IWeapon.h:80]: weapon’s

File: Gem/Code/Source/Weapons/WeaponTypes.cpp [hunk 4]
```diff
@@ -87,8 +139,30 @@ namespace MultiplayerSample
...
+                ->Field("EffectDirection", &ClientEffect::m_effectDirection);
+
+            AZ::EditContext* editContext = serializeContext->GetEditContext();
+            if (editContext)
+            {
+                editContext->Class<ClientEffect>("ClientEffect", "Parameters that control client effect spawning")
+                    ->ClassElement(AZ::Edit::ClassElements::EditorData, "")
+                    ->DataElement(AZ::Edit::UIHandlers::Default, &ClientEffect::m_effectName, "EffectName", "The effect to play upon weapon hit confirmation")
+                    ->DataElement(AZ::Edit::UIHandlers::Default, &ClientEffect::m_lifespan, "Lifespan", "The lifespan value to provide the effects manager")
+                    ->DataElement(AZ::Edit::UIHandlers::Default, &ClientEffect::m_travelToTarget, "TravelToTarget", "If true, effect will travel from origin to target position over it's lifetime")
```
Comment (AMZN-Gene) [inline Gem/Code/Source/Weapons/WeaponTypes.cpp:161]: its lifetime

File: Gem/Code/Source/Weapons/WeaponTypes.cpp [hunk 5]
```diff
@@ -132,8 +252,43 @@ namespace MultiplayerSample
...
+            if (editContext)
+            {
+                editContext->Class<WeaponParams>("WeaponParams", "Parameters that control the behaviour of a weapon")
+                    ->ClassElement(AZ::Edit::ClassElements::EditorData, "")
+                    ->DataElement(AZ::Edit::UIHandlers::ComboBox, &WeaponParams::m_weaponType, "WeaponType", "The basic type of weapon")
+                    ->DataElement(AZ::Edit::UIHandlers::Default, &WeaponParams::m_cooldownTimeMs, "CooldownTimeMs", "The number of milliseconds needed before the weapon can activate again")
+                    ->DataElement(AZ::Edit::UIHandlers::ComboBox, &WeaponParams::m_animFlag, "AnimFlag", "The animation flag to raise on the character when starting a fire sequence")
+                    ->DataElement(AZ::Edit::UIHandlers::Default, &WeaponParams::m_activateFx, "ActivateFx", "The effect to play upon weapon activation")
+                    ->DataElement(AZ::Edit::UIHandlers::Default, &WeaponParams::m_impactFx, "ImpactFx", "The effect to play at the point of impact upon weapon hit. Played predictively for autonomous clients, and authoritatively for simulated clients")
```
Comment (AMZN-Gene) [inline Gem/Code/Source/Weapons/WeaponTypes.cpp:283]: +1 for good descriptions ^_^

File: Gem/Code/Source/Weapons/WeaponTypes.h [hunk 6]
```diff
@@ -68,13 +68,19 @@ namespace MultiplayerSample
...
     //! Parameters that control client effect spawning.
     struct ClientEffect
     {
-        Multiplayer::LongNetworkString m_effectName;               // The effect to play upon weapon hit confirmation
-        float m_lifespan = 1.0f;                                   // The lifespan value to provide the effects manager
-        bool m_travelToTarget = false;                             // If true, effect will travel from origin to target position over it's lifetime
+        AZ_RTTI(ClientEffect, "{B0B4E78C-51EC-4103-BC57-4C54ED36E3DB}");
+
+        AssetStringType m_effectName;  // The effect to play upon weapon hit confirmation
+        float m_lifespan = 1.0f;       // The lifespan value to provide the effects manager
```
Comment (AMZN-Gene) [inline Gem/Code/Source/Weapons/WeaponTypes.h:82]: its

**Review by AMZN-Gene — APPROVED** [review]

**Review by AMZN-puvvadar — APPROVED** [review]
Looks good so far! It'd be a good idea to make issues for the TODOs outlined so they're not forgotten.

**Pull request merged**